src/modules/samples/cpp/src/so/SampleModule.cpp

```diff
@@ -0,0 +1,270 @@
+    int status = MMI_OK;
+    MMI_HANDLE handle = nullptr;
+
+    ScopeGuard sg{[&]()
+    {
+        if (MMI_OK == status)
+        {
+            OsConfigLogInfo(SampleLog::Get(), "MmiOpen(%s, %d) returned: %p, status: %d", clientName, maxPayloadSizeBytes, handle, status);
+        }
+        else
+        {
+            OsConfigLogError(SampleLog::Get(), "MmiOpen(%s, %d) returned: %p, status: %d", clientName, maxPayloadSizeBytes, handle, status);
+        }
+    }};
+
+    if (nullptr != clientName)
+    {
+        try
+        {
+            // Create an instance of the Sample class to be returned as an MMI_HANDLE for each client session
+            Sample* session = new (std::nothrow) Sample(maxPayloadSizeBytes);
+            if (nullptr == session)
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiOpen failed to allocate memory");
+                status = ENOMEM;
+            }
+            else
+            {
+                handle = reinterpret_cast<MMI_HANDLE>(session);
+            }
+        }
+        catch (std::exception& e)
+        {
+            OsConfigLogError(SampleLog::Get(), "MmiOpen exception thrown: %s", e.what());
+            status = EINTR;
+        }
+    }
+    else
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiOpen called with null clientName");
+        status = EINVAL;
+    }
+
+    return handle;
+}
+
+void MmiClose(MMI_HANDLE clientSession)
+{
+    Sample* session = reinterpret_cast<Sample*>(clientSession);
+    if (nullptr != session)
+    {
+        delete session;
+    }
+}
+
+int MmiSet(
+    MMI_HANDLE clientSession,
+    const char* componentName,
+    const char* objectName,
+    const MMI_JSON_STRING payload,
+    const int payloadSizeBytes)
+{
+    int status = MMI_OK;
+    Sample* session = nullptr;
+
+    ScopeGuard sg{[&]()
+    {
+        if (MMI_OK == status)
+        {
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogInfo(SampleLog::Get(), "MmiSet(%p, %s, %s, %.*s, %d) returned %d", clientSession, componentName, objectName, payloadSizeBytes, payload, payloadSizeBytes, status);
+            }
+        }
+        else
+        {
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiSet(%p, %s, %s, %.*s, %d) returned %d", clientSession, componentName, objectName, payloadSizeBytes, payload, payloadSizeBytes, status);
+            }
+            else
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiSet(%p, %s, %s, -, %d) returned %d", clientSession, componentName, objectName, payloadSizeBytes, status);
+            }
+        }
+    }};
+
+    if (nullptr == clientSession)
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiSet called with null clientSession");
+        status = EINVAL;
+    }
+    else
+    {
+        session = reinterpret_cast<Sample*>(clientSession);
+        status = session->Set(componentName, objectName, payload, payloadSizeBytes);
```

Comment:
>>Set(comp [](http://example.com/codeflow?start=25&length=9)

Pass into Sample::Set also the MMI_HANDLE clientSession so that Set validates it against its own 'this' pointer as the ultimate test that this is a valid object.